File: built-in plugins dsp/mb-creative-freqshift.cpp
/**
 * MB Frequency Shifter
 * Category : effect
 * Type     : eq
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Linear frequency shifting for metallic and inharmonic effects
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_CREATIVE_FREQSHIFT_H
#define MB_CREATIVE_FREQSHIFT_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbCreativeFreqshift : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-creative-freqshift";
    static constexpr const char* PLUGIN_NAME    = "MB Frequency Shifter";
    static constexpr const char* PLUGIN_TYPE    = "eq";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float shift = 0f;  // range [-2000, 2000]
    float feedback = 0f;  // range [0, 0.9]
    float direction = 0f;  // range [0, 2]
    float mix = 1f;  // range [0, 1]
    };

    MbCreativeFreqshift() = default;
    ~MbCreativeFreqshift() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.shift = std::clamp(params.shift, -2000f, 2000f);
        params.feedback = std::clamp(params.feedback, 0f, 0.9f);
        params.direction = std::clamp(params.direction, 0f, 2f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Frequency Shifter
        return input;
    }
};

#endif // MB_CREATIVE_FREQSHIFT_H

File: built-in plugins dsp/mb-creative-freqshift.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbCreativeFreqshiftPlugin: PluginDefinition = { id: 'mb-creative-freqshift', slug: 'mb-creative-freqshift', name: 'MB Frequency Shifter', category: 'effect', type: 'eq' as any, version: '1.0.0', description: 'Linear frequency shifting for metallic and inharmonic effects', author: 'Max Booster', parameters: [{ id: 'shift', name: 'Shift', type: 'float', defaultValue: 0, minValue: -2000, maxValue: 2000, automatable: true }, { id: 'feedback', name: 'Feedback', type: 'float', defaultValue: 0, minValue: 0, maxValue: 0.9, automatable: true }, { id: 'direction', name: 'Direction', type: 'float', defaultValue: 0, minValue: 0, maxValue: 2, automatable: false }, { id: 'mix', name: 'Mix', type: 'float', defaultValue: 1, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { shift: 0, feedback: 0, direction: 0, mix: 1 } };

export default MbCreativeFreqshiftPlugin;
